### PR TITLE
storage: allow multiple in-flight snapshots

### PR DESCRIPTION
**Pull request by tamird**
The specific limit is not currently tested - consider this out for early review.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7111)

### CODE REVIEW

**Comment by tschottdorf** [comment]
looks good in spirit, but acquiring the store lock in those raftstorage methods makes me very nervous. I wasn't able to justify that from the comments in the header of that file.

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> #### storage: allow multiple in\-flight snapshots
> 
> The specific limit is not currently tested - consider this out for early review.
> 
> 
> 
</details>


---

Reviewed 1 of 1 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJljwx1iPMWOz8TH6r1:-KJljwx1iPMWOz8TH6r2:-655227879) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/replica_raftstorage.go#L40)):*
> ```Go
> var _ raft.Storage = (*Replica)(nil)
> 
> // All calls to raft.RawNode require that an exclusive lock is held.
> ```

While you're here, could you clarify the locking comment? Looks like in the code of yours below, you assume that the store lock is held. I don't see anything about that here.

---

*[storage/replica_raftstorage.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJllNu6cSzVkXUclH9N:-KJllNu75A2Id1o_NDab:159344102) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/replica_raftstorage.go#L399)):*
> ```Go
> 
> 	cleanupFn := func() {
> 		r.store.mu.Lock()
> ```

Why lock here but not above?

---

*[storage/replica_test.go, line 5515 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJlm1WWRT_zkWNqDTpb:-KJlm1WX2hxf6YBccRv4:-1213465395) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/replica_test.go#L5515)):*
> ```Go
> 		}
> 	} else {
> 		t.Fatalf("expected snapshot channel for replica %d to exist but got %v instead", tc.rng.mu.replicaID, tc.rng.store.mu.snapshots)
> ```

nit: "needs" the lock

---

*[storage/store.go, line 414 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJlmFxbYzbjmzF0wdqE:-KJlmFxbYzbjmzF0wdqF:1795944789) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/store.go#L414)):*
> ```Go
> 	// permitted to proceed concurrently. Snapshots count against this limit from
> 	// the start of their generation until they are either discarded or sent on
> 	// the wire.
> ```

What's the point of keeping these fields unexported? Setting any of these values from the outside (usually server) required them to be exported, and we may want to display this information in debug pages, etc as it is the configuration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-:-KJlp8W4UcPda8HqpPJ1:1952617353)*

**Comment by tamird** [comment]
I didn't understand all your comments, but hopefully I've addressed enough of them?

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> looks good in spirit, but acquiring the store lock in those raftstorage methods makes me very nervous. I wasn't able to justify that from the comments in the header of that file.
</details>


---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/replica_raftstorage.go, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJljwx1iPMWOz8TH6r1:-KJmC8AdIctA5XKb0o_n:607656718) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/replica_raftstorage.go#L40)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> While you're here, could you clarify the locking comment? Looks like in the code of yours below, you assume that the store lock is held. I don't see anything about that here.
</details>
Not sure what you mean exactly - I'm not interacting with raft.RawNode in this change at all.

---

*[storage/replica_raftstorage.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJllNu6cSzVkXUclH9N:-KJmCDvFas9kScEekHZ9:-433115236) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/replica_raftstorage.go#L399)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why lock here but not above?
</details>
Good catch - this lock should only be in the async part. The reason is that this method (`Snapshot()`) assumes that the lock is held.

---

*[storage/replica_test.go, line 5515 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJlm1WWRT_zkWNqDTpb:-KJmCc7Tp4cDxzB3QeBJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/replica_test.go#L5515)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> nit: "needs" the lock
</details>
Done.

---

*[storage/store.go, line 414 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJlmFxbYzbjmzF0wdqE:-KJmCd2GHmEl2kZUDC6r:-983260773) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/store.go#L414)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's the point of keeping these fields unexported? Setting any of these values from the outside (usually server) required them to be exported, and we may want to display this information in debug pages, etc as it is the configuration.
</details>
Just instinct - not exporting stuff until it's needed. It's not like it's hard to export it later when needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-:-KJmDDHEhJaOt3Bf2t-h:-538701289)*

**Comment by tamird** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/replica_raftstorage.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJllNu6cSzVkXUclH9N:-KJmL0GQ54jAImOSOy3P:324929975) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f49d3ad1491ea692b0ddb317b3bfa5cf8b0b19a/storage/replica_raftstorage.go#L399)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Good catch - this lock should only be in the async part. The reason is that this method (`Snapshot()`) assumes that the lock is held.
</details>
Ah, I see what you meant and fixed now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111)*

**Comment by tschottdorf** [comment]
:lgtm:

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I didn't understand all your comments, but hopefully I've addressed enough of them?
</details>


---

Reviewed 5 of 5 files at r3, 6 of 6 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica_test.go, line 5509 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJqQ-mKIq0QS1Refcmr:-KJqQ-mLSsIAnHdN0zvs:-1323288485) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0b729a90ff5904b4a8a2c07434d7cfd54c231c/storage/replica_test.go#L5509)):*
> ```Go
> 	if ok {
> 		// Wait for the snapshot to be generated and abandoned.
> 		time.Sleep(100 * time.Millisecond)
> ```

Shouldn't you be able to poll for `len(r.store.snapMu.snapshots) == 0`?

---

*[storage/store.go, line 331 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJqQTCU6DdbWLs4zZiX:-KJqQTCU6DdbWLs4zZiY:825094880) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0b729a90ff5904b4a8a2c07434d7cfd54c231c/storage/store.go#L331)):*
> ```Go
> 		// Values are set to an open channel while a snapshot is being generated
> 		// for the given key (range ID). Keys for which a snapshot is not in
> 		// progress are either not present, or have values which are closed
> ```

I don't think you can currently find a value that has a closed channel since you delete first, then close.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-:-KJqQopRoY7DAOJJZrOm:1417088721)*

**Comment by tamird** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica_test.go, line 5509 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJqQ-mKIq0QS1Refcmr:-KJqZtQE2a2_1gs3ZDXc:-591771140) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0b729a90ff5904b4a8a2c07434d7cfd54c231c/storage/replica_test.go#L5509)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Shouldn't you be able to poll for `len(r.store.snapMu.snapshots) == 0`?
</details>
Yeah, something like that. Done.

---

*[storage/store.go, line 331 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111#-KJqQTCU6DdbWLs4zZiX:-KJqZkakz4QM-QoMLSQU:-1039939766) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0b729a90ff5904b4a8a2c07434d7cfd54c231c/storage/store.go#L331)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't think you can currently find a value that has a closed channel since you delete first, then close.
</details>
Yeah, that's true, but I'm not sure I want to tighten this contract.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111)*

**Comment by tschottdorf** [comment]
Reviewed 6 of 6 files at r6, 6 of 6 files at r7.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7111)*

File: storage/store.go [hunk 1]
```diff
@@ -657,6 +669,7 @@ func (sc *StoreContext) setDefaults() {
 	}
 
 	raftElectionTimeout := time.Duration(sc.RaftElectionTimeoutTicks) * sc.RaftTickInterval
+	sc.concurrentSnapshotLimit = 5
```
Comment (petermattis) [inline storage/store.go]: Is there a rationale behind this setting? I would have guessed that we'd want this set to `1`.

**Comment by petermattis** [comment]
Let's revert this change until we figure out the rebalancing issues it seems to be causing.

File: storage/store.go (hunk 1 above)
Comment (tamird) [inline storage/store.go]: no particular rationale

Comment (petermattis) [inline storage/store.go]: I'd like to see this set to 1 until we can measure an improvement from allowing concurrent snapshots.

Comment (tamird) [inline storage/store.go]: https://github.com/cockroachdb/cockroach/pull/7229 for nwo

**Pull request merged**